Lb/nncf integration

LeonidBeynenson opened the pull request.

LeonidBeynenson committed:
First implementation of NNCF integration

WIP

LeonidBeynenson committed:
Make NNCF training work, but convergence is still unsufficient

LeonidBeynenson committed:
Add freezing aux models during NNCF training

LeonidBeynenson committed:
Improve logging -- add timestamp to log lines

LeonidBeynenson committed:
Add possiility to turn off mutual learning or freeze the aux model up to some epoch

LeonidBeynenson committed:
Add debug logs in WarmupScheduler

LeonidBeynenson committed:
Move nncf stuff to a separate file

Almost debugged

LeonidBeynenson committed:
Add loading NNCF-compressed model weights

LeonidBeynenson committed:
Add NNCF to convert_to_onnx.py script

LeonidBeynenson committed:
Merge remote-tracking branch 'origin/ote' into lb/nncf_integration

LeonidBeynenson committed:
Merge branch 'ote' into lb/nncf_integration

LeonidBeynenson committed:
Add possibility to turn off NNCF in MobileNetV3 SELayer-s and head (now turned off by default)

LeonidBeynenson committed:
Make NNCF quantize SE-blocks and head of MobilenetV3 by default

LeonidBeynenson committed:
Make tools/main.py turn off NNCF compression by default

Earlier in this branch NNCF compression was turned ON by default
(for developing purposes),
so now some of the previous experiments cannot be repeated without
addition of `--nncf` flag.

Also insert several TODO-s and remove several debug prints.

LeonidBeynenson committed:
Remove debug changes to check if a model is changed during training

It was required to check if not-BN layers are changed if LR==0.

LeonidBeynenson committed:
Merge branch 'kp/classification_exp' into lb/nncf_integration

LeonidBeynenson committed:
Merge branch 'kp/classification_exp' into lb/nncf_integration

LeonidBeynenson committed:
Add minor comments

LeonidBeynenson committed:
Add possibility to modify aux config options using command-line args

LeonidBeynenson committed:
Add writing/loading NNCF config from checkpoint's meta

Also remove unrequired cmd line params.

LeonidBeynenson committed:
Fix minor bugs in add_nncf_metainfo_to_checkpoint.py

LeonidBeynenson committed:
Fix DEBUG insertions

LeonidBeynenson committed:
Merge branch 'ote' into lb/nncf_integration

LeonidBeynenson committed:
Turn on NNCF mode in convert_to_onnx.py if checkpoint has NNCF metainfo

LeonidBeynenson committed:
Decorative changes

LeonidBeynenson committed:
Add calculating initial LR for NNCF and reading NNCF config from model config

LeonidBeynenson committed:
Fix nncf part of model config

Also move initialization of optimizer and scheduler in main.py after
NNCF part of the code to use the changes in config that are done by
this NNCF part.

LeonidBeynenson committed:
Rename variable lr -> aux_lr in main.py

LeonidBeynenson committed:
Refactor tools/main.py; add NNCF part to tools/eval.py

LeonidBeynenson committed:
Fix bugs in tools/eval.py

LeonidBeynenson commented:
2 @kprokofi , @sovrasov : Kirill, Vlad, could you review this PR?
I think I made all the required changes now.

sovrasov commented on tools/auxiliary/add_nncf_metainfo_to_checkpoint.py:
```diff
@@ -0,0 +1,77 @@
+
+    # default nncf config
+    nncf_config_data = {
+#            "input_info": {
```
Why these lines are commented?

LeonidBeynenson replied:
Fixed -- see the new function `get_default_nncf_compression_config`

kprokofi commented on tools/main.py:
```diff
@@ -58,20 +135,38 @@ def main():
 
+    if cfg.model.resume and check_isfile(cfg.model.resume):
+        cfg.train.start_epoch = resume_from_checkpoint(
+            cfg.model.resume, model, optimizer=optimizer, scheduler=scheduler
```
In this part of the code, we don't have access to the optimizer and scheduler. Please, consider to move it back

LeonidBeynenson replied:
Oh, I see, sorry. I will make a fix.

LeonidBeynenson replied:
Fixed

kprokofi commented on tools/main.py:
```diff
@@ -14,19 +15,93 @@
+        # just skipping loading special datamanager
+        datamanager_for_nncf = None
+    else:
+        print('before building datamanager for nncf initializing')
```
I guess we don't need to print these messages for users

LeonidBeynenson replied:
Sorry, but I think it is important.
1. The loading of dataset may require some time, so IMHO it is a good idea to inform a user what is making now.
2. Also (and it is even more important) when a datamanager is created, some logs will be printed to output. It may be confusing for users. IMHO it is a good idea to inform user why one more datamanager is created.

LeonidBeynenson replied:
NB: this function has been moved to `torchreid/integration/nncf/compression_script_utils.py`

kprokofi commented on tools/main.py:
```diff
@@ -14,19 +15,93 @@
+                                   is_initial_lr_set_from_opts):
+    if is_initial_lr_set_from_opts:
+        print(f'Since initial LR is set from command line arguments, do not calculate initial LR for NNCF '
+              f'cfg.train.lr={cfg.train.lr}')
```
Maybe it's better to print a message about the learning rate in a different way, not like in the config file, in my opinion, it would be more understandable for users

LeonidBeynenson replied:
Ok, I will change it a bit

LeonidBeynenson replied:
Changed

kprokofi commented on tools/main.py:
```diff
@@ -14,19 +15,93 @@
+    print(f'Received from wrapping nncf_metainfo =\n{pformat(nncf_metainfo)}')
+    if cfg.lr_finder.enable:
+        print('Turn off LR finder -- it should not be used together with NNCF compression')
+        cfg.lr_finder.enable = False
```
I suggest moving this reinitializing of the learning rate to related function in `scripts/script_utils` , `reset_config`  depending on the according parameter about nncf in config file or in args.nncf

LeonidBeynenson replied:
Kirill, I think you mean not the code just above, but the code just below -- where the function `calculate_lr_for_nncf_training` is called, don't you?

Are you sure that the implementation of the function should be moved?
At the moment I think that the `tools/main.py` is the best place for the functions, since it is the only place where it may be used -- it is required for training only.

But if you are sure that the main launching script should be "as short as possible", I will move it to `torchreid/integration/nncf/compression.py`

sovrasov replied:
@LeonidBeynenson we moved most of the functions from main.py to make them a part of public API which is required to implement OTE task interface.

LeonidBeynenson replied:
Ok. I see. This is a good reason.

LeonidBeynenson replied:
Moved

kprokofi commented on tools/main.py:
```diff
@@ -115,26 +210,28 @@ def main():
+        if num_devices > 1:
+            model = DataParallel(model, device_ids=main_device_ids, output_device=0).cuda(main_device_ids[0])
+        else:
+            model = model.cuda(main_device_ids[0])
```
all of the models had been working in the data-parallel regime before, but you propose to change this behavior for 1 device. Is it necessary for NNCF? Because I can't be one hundred sure that nothing would be broken

LeonidBeynenson replied:
As far as I can see no reason to use `DataParallel` if we use one GPU device, whereas `DataParallel` prints some warnings in this case.
I checked this case, everything works ok.

kprokofi commented on tools/main.py:
```diff
@@ -14,19 +15,93 @@
 
+from torchreid.integration.nncf.compression import wrap_nncf_model, is_checkpoint_nncf
+
+def is_lr_set_from_opts(opts):
```
Please, consider to move all of the aux function for NNCF to separate utils file. For example, in `scripts/scripts_utils`. We did it for all aux functions in `main` to unload the main launching script

LeonidBeynenson replied:
Ok, I move it

LeonidBeynenson replied:
Moved

kprokofi commented on tools/eval.py:
```diff
@@ -29,8 +29,18 @@
+from torchreid.integration.nncf.compression import wrap_nncf_model, is_checkpoint_nncf
 
 
+def make_nncf_changes_in_eval(model, cfg):
```
It's also better to move in utils file, IMO

LeonidBeynenson replied:
Kirill, are you sure? If I move this function, the part of logic will be hidden here.
IMHO it is better to keep the functions `make_nncf_changes_in_training` in `tools/main.py` and the function `make_nncf_changes_in_eval` in `tools/eval.py`

I think I will move the function `calculate_lr_for_nncf_training`, since it may be isolated from other code, but I would prefer to keep `make_nncf_changes_in_training`.

LeonidBeynenson replied:
After comments @sovrasov moved it to a new file.

kprokofi commented on torchreid/engine/engine.py:
```diff
@@ -19,6 +20,33 @@
+        raise RuntimeError(f'Wrong epoch_interval {epoch_interval}')
+
+    if epoch_interval.first is not None and epoch < epoch_interval.first:
+        print(f'_get_cur_action_from_epoch_interval: since epoch={epoch} < epoch_interval.first={epoch_interval.first}, '
```
Do we actually need these prints? For users it's meaningless information, I suppose.

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

sovrasov commented on tools/main.py:
```diff
@@ -58,20 +135,38 @@ def main():
 
+    if cfg.model.resume and check_isfile(cfg.model.resume):
+        cfg.train.start_epoch = resume_from_checkpoint(
+            cfg.model.resume, model, optimizer=optimizer, scheduler=scheduler
```
Optimizer and scheduler should be created in advance in case of resume

LeonidBeynenson replied:
Ok, fixing it

LeonidBeynenson replied:
Fixed

sovrasov commented on tools/main.py:
```diff
@@ -58,20 +135,38 @@ def main():
+                                                                          args.classes,
+                                                                          is_initial_lr_set_from_opts)
+
+        is_nncf_used = True
```
Seems like this var is not used

LeonidBeynenson replied:
Now it will be used -- see my latest changes

kprokofi commented on torchreid/engine/engine.py:
```diff
@@ -19,6 +20,33 @@
+    print(f'_get_cur_action_from_epoch_interval: return value_inside={epoch_interval.value_inside}')
+    return epoch_interval.value_inside
+
+def get_initial_lr_from_checkpoint(filename):
```
I think it's an inappropriate place for this function. It's not related to Engine

LeonidBeynenson replied:
Kirill, it is really related to the method how `initial_lr` is stored in the checkpoint.
But this is defined in one place in the code -- in the function `save_model` in `Engine`
I think it is a really wrong idea to separate the method of writing `initial_lr` field to checkpoint and the method of loading it from checkpoint.
(Otherwise later when somebody change saving it could forget to change loading, and NNCF pipeline will be broken).
I could make it a static function of `Engine` next to the method `save_model` to make it more clear.

kprokofi commented on torchreid/optim/lr_scheduler.py:
```diff
@@ -192,15 +192,19 @@ def get_lr(self):
                 if not self.finished:
                     self.after_scheduler.base_lrs = [base_lr * self.multiplier for base_lr in self.base_lrs]
                     self.finished = True
+                    print(f'WarmupScheduler.get_lr: set finished=True, self.after_scheduler.base_lrs={self.after_scheduler.base_lrs}')
```
is it debug print?

LeonidBeynenson replied:
Yes, I will remove it

LeonidBeynenson replied:
Removed

kprokofi commented on torchreid/utils/torchtools.py:
```diff
@@ -307,6 +318,20 @@ def load_pretrained_weights(model, file_path='', pretrained_dict=None):
             '(** ignored and continue **)'.format(message)
         )
+
+        if len(discarded_layers) > 0:
```
I can't understand why we need to know here about discarded layers and matched layers below if we don't have matched layers at all

LeonidBeynenson replied:
It will make much more convenient comparison between the model the state dict and the state dict loaded from a checkpoint.

E.g. before I added `_remove_prefix(k, 'nncf_module')` the weights from nncf checkpoint did not load to a model before it is wrapped by NNCF, whereas it is required as the initial stage of NNCF wrapping. 
The prints just showed what happened (`nncf_module` prefix)
Without the prints a user will have to do the same manually.

kprokofi commented on torchreid/utils/torchtools.py:
```diff
@@ -307,6 +318,20 @@ def load_pretrained_weights(model, file_path='', pretrained_dict=None):
             'please check the key names manually '
             '(** ignored and continue **)'.format(message)
```
since now we raise an error, the comment `** ignored and continue **` is not needed anymore.

LeonidBeynenson replied:
Ok, I will remove it

LeonidBeynenson replied:
Removed

sovrasov commented on torchreid/utils/torchtools.py:
```diff
@@ -307,6 +318,20 @@ def load_pretrained_weights(model, file_path='', pretrained_dict=None):
             '(** ignored and continue **)'.format(message)
         )
+
+        if len(discarded_layers) > 0:
```
There is a duplication of discarded and unmatched layer names printing: code in this and the next branch is identical. May de move if to a `_*` function?

LeonidBeynenson replied:
Ok, I will do it

LeonidBeynenson replied:
Fixed

sovrasov reviewed (commented):
First bunch of comments

sovrasov commented on torchreid/integration/nncf/compression.py:
```diff
@@ -0,0 +1,247 @@
+                           f'datamanager_for_init={datamanager_for_init} checkpoint_path={checkpoint_path}')
+    if not nncf_config_data:
+        print('Using the default NNCF int8 quantization config')
+        nncf_config_data = {
```
Seems like the same structure appears in `add_nncf_metainfo_to_checkpoint.py`. Could we keep it here and import in other files?

LeonidBeynenson replied:
Ok, I will move it to a special file.

LeonidBeynenson replied:
Moved it to a separate function `get_default_nncf_compression_config` in `torchreid/integration/nncf/compression.py`

sovrasov commented on torchreid/integration/nncf/compression.py:
```diff
@@ -0,0 +1,247 @@
+        norm_mean=cfg.data.norm_mean,
+        norm_std=cfg.data.norm_std,
+    )
+    def random_image(height, width):
```
`random_image` was already defined in `scripts.script_utils`

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed.
Please, note that to use `random_image` in `torchreid.integration.nncf.compression` I had to move it to `torchreid.utils.tools`, otherwise cyclic import issue happened.

sovrasov commented on torchreid/integration/nncf/compression.py:
```diff
@@ -0,0 +1,247 @@
+
+        return out_img
+
+    def dummy_forward(model):
```
Could we use this function in `convert_to_onnx.py` as well?

LeonidBeynenson committed:
Refactor code, move some code to new file compression_script_utils.py

Also remove some debug prints.

LeonidBeynenson committed:
Refactor printing info on unmatched/discarded layers in load_pretrained_weights

LeonidBeynenson committed:
Fix bug: resume_from_checkpoint should be after creation of scheduler and optimizer

So, move initial loading of the model for NNCF into
make_nncf_changes_in_training.

LeonidBeynenson commented:
@kprokofi , @sovrasov : made fixes w.r.t. most of your comments.
Please, take a look

LeonidBeynenson committed:
Move random_image to torchreid.utils.tools; implement function get_default_nncf_compression_config

Note that random_image should be in torchreid.utils.tools to use it in
torchreid.integration.nncf.compression, otherwise cyclic import happens.

LeonidBeynenson commented:
2 @kprokofi, @sovrasov : Kirill, Vlad, now I made changes w.r.t. all your comments (added one more commit from yesterday).

Please, re-review the code, write me if everything is OK now.

sovrasov commented on tools/main.py:
```diff
@@ -2,6 +2,7 @@
 import os.path as osp
 import sys
 import time
+from pprint import pformat
```
Seems like `pformat` is not used in this file

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Add possibility to change main/aux configs for NNCF training

Also make some refactoring.

LeonidBeynenson commented:
2 @kprokofi , @sovrasov : Kirill, Vlad, I added possibility to change the main and aux config files for NNCF training.

Also some refactoring is made.

Could you re-review my changes and accept them if everything is OK?

kprokofi commented on torchreid/integration/nncf/compression.py:
```diff
@@ -0,0 +1,233 @@
+        model(input_blob)
+        model.train(prev_training_state)
+
+    # TODO: think if this is required
```
Have you already solved your TODO? Or it's not important?

LeonidBeynenson replied:
Removed one of TODO, reworded the other one.

Now I don't want to change this part of code any more.

kprokofi commented:
@LeonidBeynenson, I have one general question about EfficientNet_b0. Won't it be wrapped in NNCF?

LeonidBeynenson committed:
Fix TODO-s in the code

LeonidBeynenson commented:
> @LeonidBeynenson, I have one general question about EfficientNet_b0. Won't it be wrapped in NNCF?

I have not started work on it. I hope I will do it a bit later.
Anyway, I'm not going to add its support to this PR. )
